src/katgpucbf/xbgpu/kernels/beamform.mako

```diff
@@ -50,65 +52,96 @@ DEVICE_FN float2 cmul(float2 a, float2 b)
 // Compute a * b + c in complex numbers
 DEVICE_FN float2 cmad(float2 a, float2 b, float2 c)
 {
-    return make_float2(a.x * b.x - a.y * b.y + c.x, a.x * b.y + a.y * b.x + c.y);
+    return make_float2(c.x + a.x * b.x - a.y * b.y, c.y + a.x * b.y + a.y * b.x);
```

Comment:
I see this change belongs to an optimisation commit with the blurb
> Reorder operations in cmad to use **fused operations throughout** [emphasis added]

My (admittedly few minutes of) Google'ing didn't return anything overly useful, at least on first read. Basically,
* What is a fused operation? And/or
* Why is this update better than it was before?

Reply:
Fused multiply-add (FMA): a single instruction that computes x*y+z (with correct rounding, rather than rounding x*y first). NVIDIA GPUs execute that twice as fast as using separate multiple and add instructions. The compiler was previously generating a MUL, and FMA and an ADD (I forget in what order). With the new version it is two FMA instructions.

Reply:
Two points
First, Re:
> The compiler was previously generating a MUL, an FMA and an ADD (I forget in what order)

Not sure how else to word "How did you know that?".  I guess, now that you've pointed it out, I perhaps see 
* `a.x * b.x` as the single MUL instruction,
* `a.x * b.y + c.x` as an FMA,
* and the leftover being the ADD.

Whereas now, with your update
* The first FMA == `c.x + a.x * b.x` , and
* ... I'm actually not sure how the second FMA comes about? Is there still a final ADD?

Apologies if this is ~incoherent, happy to Zoom for a few mins.

Reply:
> How did you know that?

Nsight showed me the instruction mix (which had lots of FMUL and FADD) and I think maybe even warned me about the low density of FFMA.

> ... I'm actually not sure how the second FMA comes about? Is there still a final ADD?

Let t = c.x + a.x * b.x (the first FMA). Then the second FMA is t - a.y * b.y (it's a subtract rather than an add, but IIRC NVIDIA lets you negate operands for free).

Reply:
Ah, fantastic. I was guesstimating the result of `a.y * b.y` gets folded into another FMA, but it's actually/likely the former that gets computed first (maybe due to left-to-right execution?).

Either way, explanation is much appreciated.